### PR TITLE
rapid-photo-downloader: 0.9.13 -> 0.9.14

### LINKED ISSUE #56815: rapid-photo-downloader / rawkit: fails to build
## Issue description

Rapid Photo Downloader depends on the Python library rawkit for bindings to libraw.

Rawkit fails to build due to test failures.

### Steps to reproduce

1. nix-shell -p rapid-photo-downloader

Rawkit's check phase blows up with:

```
[...]
tests/unit/options_test.py:62: RuntimeError
===================== 1 failed, 52 passed in 0.30 seconds ======================
builder for '/nix/store/zqgarbyc2f930h3gdw438zpd87nn6abv-python3.7-rawkit-0.6.0.drv' failed with exit code 1
cannot build derivation '/nix/store/z6hljn124mjnry4iigja9mcj4jfjwpi4-rapid-photo-downloader-0.9.13.drv': 1 dependencies couldn't be built
error: build of '/nix/store/z6hljn124mjnry4iigja9mcj4jfjwpi4-rapid-photo-downloader-0.9.13.drv' failed
```

## Proposed Fix

Rawkit is unmaintained, but has merged fixes into its master branch since its last release.

Using the current `master` branch of https://github.com/photoshell/rawkit works.

Note, rawkit's master branch also requires adding the Python package `numpy` to `buildInputs`

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

 - system: `"x86_64-linux"`
 - host os: `Linux 4.20.13, NixOS, 19.09.git.80812af (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.09pre170705.1233c8d9e9b"`
 - channels(dan): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by callahad**
This updates Rapid Photo Downloader to 0.9.14 and fixes #56815

###### Motivation for this change

- Version bump
- Actually make it work:
    - Does not require `exiftool` to be present in the user environment
    - Can now successfully find `libmediainfo`
    - Can now successfully generate video thumbnails

Known issues:

- Support for RAW images is not fully functional.

    [Rapid Photo Downloader](https://www.damonlynch.net/rapid/) uses [Rawkit](https://rawkit.readthedocs.io/en/latest/) bindings to [libraw](https://www.libraw.org/).

    Rawkit is unmaintained, and its most recent release does not support libraw 0.19. The nixpkgs collection has included 0.19 since [mid-2018](https://github.com/NixOS/nixpkgs/commit/92d48c6a21a0fe1c814d611b8a923688404f83f0#diff-0d9f1728501ca545bce47c7ed02e1214).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by callahad** [comment]
@jfrankenau, as maintainer of rapid-photo-downloader, I'd appreciate your review.

@dotlambda, you helped out with unblocking #56815, your review would also be appreciated.

**Comment by grahamc** [comment]
@GrahamcOfBorg build rapid-photo-downloader

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix [hunk 1]
```diff
@@ -19,24 +19,31 @@ python3Packages.buildPythonApplication rec {
...
       --replace "from preferences" "from raphodo.preferences"
-    substituteInPlace raphodo/copyfiles.py \
-      --replace "import problemnotification" "import raphodo.problemnotification"
   '';
 
-  nativeBuildInputs = [ file intltool gobject-introspection ];
+  nativeBuildInputs = [
+    exiftool
+    file
+    intltool
```
Comment (dotlambda) [inline pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Why was `gobject-introspection` removed? And why is `exiftool` there? Maybe the three packages listed are actually `checkInputs`?

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix [hunk 2]
```diff
@@ -54,11 +61,14 @@ python3Packages.buildPythonApplication rec {
...
     pyprind
-  ]);
+  ];
 
   makeWrapperArgs = [
     "--set GI_TYPELIB_PATH \"$GI_TYPELIB_PATH\""
     "--set PYTHONPATH \"$PYTHONPATH\""
+    "--prefix PATH : ${stdenv.lib.makeBinPath [ exiftool ]}"
+    "--prefix LD_LIBRARY_PATH : ${stdenv.lib.makeLibraryPath [ libmediainfo ]}"
+    "--prefix GST_PLUGIN_SYSTEM_PATH_1_0 : \"$GST_PLUGIN_SYSTEM_PATH_1_0\""
```
Comment (dotlambda) [inline pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Why is this needed? Shouldn't this be `--set`?

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix [hunk 3]
```diff
@@ -19,24 +19,31 @@ python3Packages.buildPythonApplication rec {
...
-  ] ++ (with python3Packages; [
+    gexiv2
+    gobject-introspection
+    gst_all_1.gst-libav
+    gst_all_1.gst-plugins-base
+    gst_all_1.gst-plugins-good
+    gst_all_1.gstreamer
+    gst_all_1.gstreamer.dev
+    libgudev
+    libmediainfo
```
Comment (dotlambda) [inline pkgs/applications/graphics/rapid-photo-downloader/default.nix]: `libmediainfo` is specified in `makeWrapperArgs` and can thus be removed from here

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix (hunk 1 above)
Comment (callahad) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: `intltool` and `file` are both invoked during the normal build process, even with `doCheck = false;`

But you're right, `exiftool` can be moved to `checkInputs`

Comment (callahad) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: As for `gobject-introspection`, I honestly don't understand why, but the build only succeeds if it's in in `buildInputs`. Whether or not it's in `nativeBuildInputs` doesn't seem to matter.

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix (hunk 3 above)
Comment (callahad) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Done

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix (hunk 2 above)
Comment (callahad) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Without setting `GST_PLUGIN_SYSTEM_PATH_1_0`, gstreamer can't find its plugins, and thus fails to generate thumbnails for video files.

`--set` seems reasonable to me, but virtually every other occurrence of `GST_PLUGIN_SYSTEM_PATH` in nixpkgs, including `wrapGAppsHook`, uses `--prefix`. Out of more than two dozen occurrences, only one (beets) uses `--set`.

Comment (dotlambda) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Okay, let's keep `--prefix`.

File: pkgs/applications/graphics/rapid-photo-downloader/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: Okay, no idea about why that's needed. However, please include a comment that without it we get `Namespace GUdev not available`.

Comment (callahad) [reply pkgs/applications/graphics/rapid-photo-downloader/default.nix]: ¯\\\_(ツ)\_/¯

Added a a comment noting the error.

**Review by dotlambda — CHANGES_REQUESTED** [review]
I'm seeing
> Ran 0 tests in 0.001s

Was this the case before? We should set `doCheck = false` and add a comment with the reason if there are no tests.

**Comment by callahad** [comment]
The `checkPhase` definitely does something, because the log output is different:

```diff
--- nocheck.txt 2019-04-14 19:37:14.927437916 +0100
+++ check.txt   2019-04-14 19:37:18.031466371 +0100
@@ -538,4 +538,33 @@
 wrapping `/nix/store/...-rapid-photo-downloader-0.9.14/bin/analyze-pv-structure'...
 Rewriting #!/nix/store/lmwk7mg8y79m3izdxdlckdn385x7jgl7-python3-3.7.3/bin/python3.7 to #!/nix/store/lmwk7mg8y79m3izdxdlckdn385x7jgl7-python3-3.7.3
 wrapping `/nix/store/...-rapid-photo-downloader-0.9.14/bin/rapid-photo-downloader'...
+running install tests
+running test
+running egg_info
+writing rapid_photo_downloader.egg-info/PKG-INFO
+writing dependency_links to rapid_photo_downloader.egg-info/dependency_links.txt
+writing entry points to rapid_photo_downloader.egg-info/entry_points.txt
+writing requirements to rapid_photo_downloader.egg-info/requires.txt
+writing top-level names to rapid_photo_downloader.egg-info/top_level.txt
+reading manifest file 'rapid_photo_downloader.egg-info/SOURCES.txt'
+reading manifest template 'MANIFEST.in'
+warning: no files found matching '*.txt' under directory 'raphodo'
+warning: no previously-included files matching '*~' found anywhere in distribution
+warning: no previously-included files matching '*.pyc' found anywhere in distribution
+warning: no previously-included files matching '*~1~' found anywhere in distribution
+warning: no previously-included files matching '*~2~' found anywhere in distribution
+warning: no previously-included files matching '*bak' found anywhere in distribution
+warning: no previously-included files found matching 'raphodo/test.py'
+warning: no previously-included files found matching 'raphodo/test2.py'
+no previously-included directories found matching 'tasks'
+no previously-included directories found matching 'raphodo/tests'
+no previously-included directories found matching 'debian'
+no previously-included directories found matching 'images'
+writing manifest file 'rapid_photo_downloader.egg-info/SOURCES.txt'
+running build_ext
+
+----------------------------------------------------------------------
+Ran 0 tests in 0.000s
+
+OK
 postPatchMkspecs
```

...and parts fail without `exiftool` in `checkInputs`:

```diff
--- check.txt   2019-04-14 19:37:18.031466371 +0100
+++ check-without-exiftool.txt  2019-04-14 19:38:00.453815232 +0100
@@ -562,6 +562,7 @@
 no previously-included directories found matching 'images'
 writing manifest file 'rapid_photo_downloader.egg-info/SOURCES.txt'
 running build_ext
+ERROR:root:Unable to compare ExifTool version number:
 
 ----------------------------------------------------------------------
 Ran 0 tests in 0.000s
```

But as far as I can tell, there aren't any actual unit tests in the upstream source. The program *does* contain a bunch of doctests, but they don't appear to run by default, and at least one has a comment stating that it requires actual hardware plugged in before it will pass.

...that said, I'm not seeing a meaningful difference in build time with and without `checkPhase`, so I'd be inclined to leave it as is should the author ever decide to add a proper test suite. Is there a compelling reason to disable it instead?

**Comment by dotlambda** [comment]
> But as far as I can tell, there aren't any actual unit tests in the upstream source. The program does contain a bunch of doctests, but they don't appear to run by default, and at least one has a comment stating that it requires actual hardware plugged in before it will pass.

Please run these doctests in a `checkPhase`.

**Comment by callahad** [comment]
> Please run these doctests in a checkPhase.

I don't think that makes sense. If the author intended for them to run regularly, they'd be executed as part of `setup.py test`, but they aren't. They also have things like this comment in raphodo/devices.py:

> To run the doctests, ensure at least one camera is plugged in but not mounted!

...which then goes on to test paths that hardcoded to the author's environment `/media/damon`, as well as his Canon camera `EOS_DIGITAL`.

Making these tests run and pass in the general case would require nontrivial patches against upstream.

**Comment by dotlambda** [comment]
Alright, then please set `doCheck = false` and mention the requirement of a camera as reason.

**Comment by callahad** [comment]
Done.

I still feel bad about disabling the check phase altogether as it adds trivial overhead, but if you think it's better this way, there it is. :)

**Comment by dotlambda** [comment]
Btw, if someone is interested in RAW support, they should make a PR pinning `rawkit` to 0.18 using `packageOverrides`.
